Editing product v3

EasySaber opened the pull request.

Lovkachebasher2 committed:
Product Page vol .1 

не могу добавить картинку (

Lovkachebasher2 committed:
Merge remote-tracking branch 'origin/develop' into develop

попытка 5

Lovkachebasher2 committed:
+

Lovkachebasher2 committed:
Merge remote-tracking branch 'origin/develop' into develop

вфцв

Lovkachebasher2 committed:
Merge remote-tracking branch 'origin/develop' into develop

Lovkachebasher2 committed:
Merge remote-tracking branch 'Lovkachebasher2/ProductPage' into ProductPage

Lovkachebasher2 committed:
v

Lovkachebasher2 committed:
Не видит стили

EasySaber committed:
fix: директория хранения изображений товара.

EasySaber committed:
fix: директория хранения изображений товара.

mophubajib committed:
rename about controller

mophubajib committed:
commit

EasySaber committed:
Admin: список товаров, удаление, редактирование

EasySaber committed:
Admin: список товаров, удаление, редактирование

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,138 @@
-            list = productService.findByName(productName);
-        }
-        model.addAttribute("productList", list);
+    public String saveDataForm(Model model) {
```
метод все таки надо переименовать

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,138 @@
 @AllArgsConstructor
 public class AdminProductsController {
-    private final ProductService productService;
+    private final ProductServiceImpl productService;
```
Здесь можно указать интерфейс

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,138 @@
+    private final ProductServiceImpl productService;
+    private final CategoryService categoryService;
+    private final ProductStatusService productStatusService;
+    private final ProductRepository productRepository;
```
репозиторий тут не нужен если у тебя есть сервис

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,138 @@
     @PostMapping(value="/product/add")
-    public String addProduct(Model model) {
-        return "redirect:/admin/product/add";
+    public String addProduct(@RequestParam Long category_id,
```
Одинаковые методы addProduct и editProduct, различаются только параметром productId

Тут можно сделать метод типа ```@PostMapping(value="/product/save")``` а productId передавать как необязательный ```@RequestParam```, если айдишник есть, можно достать сущность из БД обновить поля, если нет, то создать ДТО и сохранить

Можно проверить, если в ДТО проставить айдишник существующей сущности и попробовать сохранить, будет ошибка или запись в БД обновиться?

you-all-die commented on src/main/java/com/example/internship/service/impl/ProductServiceImpl.java:
```diff
@@ -42,11 +42,22 @@ public void addProduct(ProductDto productDto) {
     }
 
+    @Override
+    public List<ProductDto> findById(Long id) {
```
Зачем тут список? Айдишник же уникальный для сущности, и глядя на этот метод никак не ожидаешь что по одному айдишнику несколько продуктов может быть

you-all-die commented on src/main/resources/templates/admin/productadd.html:
```diff
@@ -0,0 +1,81 @@
+<!DOCTYPE html>
```
формы добавления и редактирования можно объединить?

Разница только в том что при добавлении из контроллера приходит пустой ProductDto а при редактировании предзаполненный, и тут при редактировании поле id частью формы становится скрытым инпутом

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,138 @@
+                              Model model) throws IOException {
+
+        //Проверка файла, создание директории, сохранение в папку /img_product/
+        if (picturefile != null) {
```
т.е. если файл  не прикрепили то и сохранить / создать нелья?

может сделать так что бы было можно, и если картинки нет, то в поле проставлять значение по умолчанию имя существующей дефолтной картинки?

gustav474 committed:
Восстановил файлы

gustav474 committed:
Восстановил файлы

gustav474 committed:
Merge pull request #76 from you-all-die/authorization

Авторизация

gustav474 committed:
Добавил CustomerRepository

gustav474 committed:
Добавил CustomerRepository

Lovkachebasher2 committed:
Merge remote-tracking branch 'origin/develop' into ProductPage

gustav474 committed:
Добавид CustomerRepository

gustav474 committed:
Добавид CustomerRepository

gustav474 committed:
Merge pull request #77 from you-all-die/authorization_fix

Authorization fix

Lovkachebasher2 committed:
Merge remote-tracking branch 'origin/develop' into ProductPage

Lovkachebasher2 committed:
Страница продукта

samyual committed:
Очередной патч несчастного главного шаблона

samyual committed:
Очередной патч несчастного главного шаблона

samyual committed:
Merge pull request #79 from you-all-die/fix-main-template-killed-by-authorization

Очередной патч несчастного главного шаблона

alexey-murashov committed:
исправил последовательность автоинкремента для id пользователей

alexey-murashov committed:
исправил последовательность автоинкремента для id пользователей

alexey-murashov committed:
Merge pull request #80 from you-all-die/customers-seq-fix

исправил последовательность автоинкремента для id пользователей

Lovkachebasher2 committed:
Merge pull request #66 from you-all-die/ProductPage

Product page

EasySaber committed:
Update.
- Админка: добавление, редактирование товара:
   - Фото сохраняется в папке /img_product/ (id_category_name.*)
   - Объединение форм добавление/редактирование
   - Объединение методов
   - Добавил default.jpg

EasySaber committed:
Update.
- Админка: добавление, редактирование товара:
   - Фото сохраняется в папке /img_product/ (id_category_name.*)
   - Объединение форм добавление/редактирование
   - Объединение методов
   - Добавил default.jpg

EasySaber committed:
Merge remote-tracking branch 'origin/editing_product_V3' into editing_product_V3

# Conflicts:
#	src/main/java/com/example/internship/service/impl/ProductServiceImpl.java

EasySaber committed:
Конфликт

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
+    private final ProductStatusService productStatusService;
 
     @GetMapping({"/products"})
-    public String saveDataForm(Model model, @RequestParam(value = "name", required = false) String productName) {
```
вот это все таки надо оставить, иначе поиск по наименованию не будет работать

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
-    public String deleteProduct(@RequestParam("productId") Long id, Model model) {
-        //здесь должны быть проверки на возможность удаления продукта, а пока просто удаляем
+    //Удаление товара
+    @GetMapping(value="/product/{productId}/delete")
```
Удаление все таки больше к POST запросу относится

Обычно в контроллерах пишут GET для методов, которые возвращают данные, и POST для методов которые меняют данные

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
-        return "redirect:/admin/product/edit?id=" + id;
-    }
 
     @PostMapping(value="/product/hide")
```
тогда и здесь можно поменять маппинг на /product/{productId}/hide что бы все единообразно было

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
+                              @RequestParam("picture_file") MultipartFile pictureNew,
+                              Model model) throws IOException {
+
+        String filePicturename = "default.jpg";
```
различные значения по умолчанию обычно выносятся в переменные класса, например
private static final String DEFAULT_FILE_NAME = "default.jpg";
и используют так 
String filePicturename = DEFAULT_FILE_NAME;

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
+    }
+
+    //Форма добавления нового продукта
+    @GetMapping(value="/addnewproduct")
```
здесь тоже стоит поправить /product/add

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
+        if (product.getId() != null){
+            if (pictureNew.getSize() > 0){
+                filePicturename = product.getId() + "_" + product.getCategory().getId() + "_" +
+                                  product.getName() + expansionFile;
```
имя товара может содержать различные символы, например тире, кавычки и т.д. и в таком случае файл может не создаться

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
+
+        String filePicturename = "default.jpg";
+        Boolean renamePicture = false;
+        String expansionFile = null;
```
тут все таки подойдет больше fileExtension )

you-all-die commented on src/main/java/com/example/internship/controller/admin/AdminProductsController.java:
```diff
@@ -1,61 +1,133 @@
+            if (renamePicture & (pictureNew.getSize() > 0)){
+                long prId = productService.findMaxProduct();
+                product = productService.findById(prId);
+                filePicturename = product.getId() + "_" + product.getCategory().getId() + "_" +
```
вот этот код повторяется несколько раз, можно вынести в приватный метод, который принимает продукт и расширение файла, и возвращает готовое имя

you-all-die commented on src/main/java/com/example/internship/controller/home/HomeControllers.java:
```diff
@@ -21,4 +21,7 @@ public String homePage(Model model) {
         return "home/index";
     }
+
+    @GetMapping("login")
```
вроде не относится к задаче

you-all-die commented on src/main/java/com/example/internship/controller/products/ProductController.java:
```diff
@@ -23,4 +24,16 @@ public String getProducts(Model model) {
+     * @Author Роман Каравашкин
+     */
+    @GetMapping("product/{id}")
+    public String showProduct(@PathVariable("id") long id, Model model){
```
не относится к задаче

EasySaber committed:
Редактирование

EasySaber committed:
fix

EasySaber committed:
Merge remote-tracking branch 'origin/editing_product_V3' into editing_product_V3
